swig/cpp/src/Libyang.cpp

```diff
@@ -0,0 +1,297 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+#include <vector>
+
+#include "Xml.hpp"
+#include "Internal.hpp"
+#include "Libyang.hpp"
+#include "Tree_Data.hpp"
+#include "Tree_Schema.hpp"
+
+extern "C" {
+#include "../../../src/libyang.h"
+#include "../../../src/tree_data.h"
+#include "../../../src/tree_schema.h"
+}
+
+using namespace std;
+
+Context::Context(ly_ctx *ctx, S_Deleter deleter) {
+    _ctx = ctx;
+    _deleter = deleter;
+}
+Context::Context(const char *search_dir, int options) {
+    _ctx = ly_ctx_new(search_dir, options);
+    if (NULL == _ctx) {
+        throw runtime_error("can not create new context");
+    }
+    _deleter = S_Deleter(new Deleter(_ctx));
+}
+Context::Context(const char *search_dir, const char *path, LYD_FORMAT format, int options) {
+    _ctx = ly_ctx_new_ylpath(search_dir, path, format, options);
+    if (NULL == _ctx) {
+        throw runtime_error("can not create new context");
+    }
+    _deleter = S_Deleter(new Deleter(_ctx));
+}
+Context::Context(const char *search_dir, LYD_FORMAT format, const char *data, int options) {
+    _ctx = ly_ctx_new_ylmem(search_dir, data, format, options);
+    if (NULL == _ctx) {
+        throw runtime_error("can not create new context");
+    }
+    _deleter = S_Deleter(new Deleter(_ctx));
+}
+Context::~Context() {}
+S_Data_Node Context::info() {
+    struct lyd_node *node = ly_ctx_info(_ctx);
+    return node ? S_Data_Node(new Data_Node(node, _deleter)) : NULL;
+}
+S_Module Context::get_module(const char *name, const char *revision, int implemented) {
+    const struct lys_module *module = ly_ctx_get_module(_ctx, name, revision, implemented);
+    return module ? S_Module(new Module((lys_module *) module, _deleter)) : NULL;
+}
+S_Module Context::get_module_older(S_Module module) {
+    const struct lys_module *new_module = ly_ctx_get_module_older(_ctx, module->_module);
+    return new_module ? S_Module(new Module((lys_module *) new_module, _deleter)) : NULL;
+}
+S_Module Context::load_module(const char *name, const char *revision) {
+    const struct lys_module *module = ly_ctx_load_module(_ctx, name, revision);
+    return module ? S_Module(new Module((lys_module *) module, _deleter)) : NULL;
+}
+S_Module Context::get_module_by_ns(const char *ns, const char *revision, int implemented) {
+    const struct lys_module *module = ly_ctx_get_module_by_ns(_ctx, ns, revision, implemented);
+    return module ? S_Module(new Module((lys_module *) module, _deleter)) : NULL;
+}
+vector<S_Module> *Context::get_module_iter() {
+    const struct lys_module *mod = NULL;
+    uint32_t i = 0;
+
+    auto s_vector = new vector<S_Module>;
+    if (NULL == s_vector) {
```

Comment:
This never happens. `operator new` never returns `nullptr` for allocations like these in C++

Reply:
ok.